### PR TITLE
Use C API for shape inference

### PR DESCRIPTION
**Pull request by malmaud**

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -226,9 +211,12 @@ function with_def_graph(ex)
     error("Improper use of with_def_graph")
     (kwargs === nothing) && (kwargs = [])
     new_args = args[2:end]
+    # Temporarily eliminate passing through keyword arguments since
+    # that breaks Revise at the moment. Should add back once Revise
+    # is patched.
```
Comment (oxinabox) [inline src/core.jl]: Can an Issue be openned to remind to fix this?
(I just stopped using Revise)

File: src/core.jl [hunk 2]
```diff
@@ -1282,7 +1273,7 @@ function setindex!(desc::NodeDescription, value::TensorShape, attr_name)
     if value.rank_unknown
         dims = Int[]
     else
-        dims = Int[(isnull(dim) ? -1 : get(dim)) for dim in value.dims]
+        dims = Int[ismissing(dim) ? -1 : dim for dim in value.dims]
```
Comment (oxinabox) [inline src/core.jl]: `dims = coalesce.(value.dims, Ref{-1})`

File: src/core.jl [hunk 3]
```diff
@@ -1750,3 +1742,60 @@ function Operation(node_def::tensorflow.NodeDef)
 
     Operation(desc)
 end
+
+function set_tensor_shape(tensor::AbstractTensor, dims)
+    output = Port(tensor)
+    g = get_graph(tensor)
+    status = Status()
+    dims = [ismissing(dim) ? -1 : dim for dim in dims]
```
Comment (oxinabox) [inline src/core.jl]: `dims = coalesce.(dims, Ref{-1})`

File: src/core.jl [hunk 4]
```diff
@@ -1750,3 +1742,60 @@ function Operation(node_def::tensorflow.NodeDef)
...
+    end
+end
+
+function get_shape(tensor::AbstractTensor, dim)
+    shape = get_shape(tensor)
+    if shape.rank_unknown
+        error("Rank is unknown")
+    end
+    if length(shape.dims) < dim
+        error("$dim is more than shape's rank of $(length(shape.dims))")
```
Comment (oxinabox) [inline src/core.jl]: This should be an `BoundsError`

File: src/core.jl (hunk 1 above)
Comment (malmaud) [reply src/core.jl]: I opened an issue on Revise and it was fixed there; I've deleted this comment on master.

File: src/core.jl (hunk 2 above)
Comment (malmaud) [reply src/core.jl]: Oh, didn't know about `coalesce`. Awesome.

**Pull request merged**